stl/inc/__msvc_int128.hpp

```diff
@@ -1323,15 +1442,111 @@ struct common_type<_Unsigned128, _Signed128> {
     using type = _Unsigned128;
 };
 
+inline namespace literals {
+    inline namespace _Int128_literals {
+        namespace _Detail {
+            enum class _U128_parse_status : unsigned char {
+                _Valid,
+                _Overflow,
+                _Invalid,
+            };
+
+            struct _U128_parse_result {
+                _U128_parse_status _Status_code;
+                _Unsigned128 _Value;
+            };
+
+            _CONSTEVAL unsigned int _Char_to_digit(const char _Ch) noexcept {
+                if (_Ch >= '0' && _Ch <= '9')
+                    return static_cast<unsigned int>(_Ch - '0');
+                if (_Ch >= 'A' && _Ch <= 'F')
+                    return static_cast<unsigned int>(_Ch - 'A' + 10);
+                if (_Ch >= 'a' && _Ch <= 'f')
+                    return static_cast<unsigned int>(_Ch - 'a' + 10);
+                return static_cast<unsigned int>(-1);
+            }
+
+            template <unsigned int _Base, char... _Chars>
+            _CONSTEVAL _U128_parse_result _Parse_u128_impl() noexcept {
+                constexpr char _Char_seq[]{_Chars..., '\0'};
+                constexpr auto _U128_max = (_STD numeric_limits<_Unsigned128>::max)();
+
+                _Unsigned128 _Val{};
+                for (size_t _Ind = 0; _Ind != sizeof...(_Chars); ++_Ind) {

```

Comment:
Why are we using a `for (index)` loop here, as opposed to a simple for-each loop?

```suggestion
                constexpr char _Char_seq[]{_Chars...};
                constexpr auto _U128_max = (_STD numeric_limits<_Unsigned128>::max)();

                _Unsigned128 _Val{};
                for (char _Char : _Char_seq) {
```

Reply:
Oh... range-`for` should be used here.

I can't simply change this to `constexpr char _Char_seq[]{_Chars...};`, because `_Chars...` is empty when writing `0__u128` (`0` is treated as the prefix of an octal integer literal).

Reply:
ugh, the lack of zero-size array types in C++ is so annoying.